addons/binding/org.openhab.binding.elkm1/src/main/java/org/openhab/binding/elkm1/internal/handler/ElkM1BridgeHandler.java

```diff
@@ -0,0 +1,375 @@
+        }
+    }
+
+    /**
+     * Called when the configuration is updated. We will reconnect to the elk at this point.
+     */
+    @Override
+    public void handleConfigurationUpdate(Map<String, Object> configurationParameters) {
+        super.handleConfigurationUpdate(configurationParameters);
+        this.connection.removeElkListener(this);
+        this.connection.shutdown();
+        this.connection = new ElkAlarmConnection(getConfigAs(ElkAlarmConfig.class), messageFactory);
+        connection.addElkListener(this);
+        if (connection.initialize()) {
+            connection.sendCommand(new Version());
+            connection.sendCommand(new ZoneDefinition());
+            connection.sendCommand(new ZonePartition());
+            connection.sendCommand(new ZoneStatus());
+            connection.sendCommand(new ArmingStatus());
+            updateStatus(ThingStatus.ONLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Requesting version from alarm");
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Unable to open socket to alarm");
+        }
+    }
+
+    /**
+     * Shutdown the bridge.
+     */
+    @Override
+    public void dispose() {
+        connection.shutdown();
+        areas = null;
+        connection = null;
+        messageFactory = null;
+        assert (listeners.isEmpty());
+        super.dispose();
+    }
+
+    /**
+     * Handlers an incoming message from the elk system.
+     *
+     * @param message The message from the elk to handle
+     */
+    @Override
+    public void handleElkMessage(ElkMessage message) {
+        logger.debug("Got Elk Message: {}", message.toString());
+        if (message instanceof VersionReply) {
+            VersionReply reply = (VersionReply) message;
+            // Set the property.
+            getThing().setProperty(ElkM1BindingConstants.PROPERTY_VERSION, reply.getElkVersion());
+            updateStatus(ThingStatus.ONLINE);
+        }
+        if (message instanceof ZoneStatusReply) {
+            ZoneStatusReply reply = (ZoneStatusReply) message;
+            for (int i = 0; i < ElkMessageFactory.MAX_ZONES; i++) {
+                Thing thing = getThingForType(ElkTypeToRequest.Zone, i + 1);
+                if (thing != null) {
+                    ElkM1ZoneHandler handler = (ElkM1ZoneHandler) thing.getHandler();
+                    if (handler != null) {
+                        handler.updateZoneConfig(reply.getConfig()[i], reply.getStatus()[i]);
+                    }
+                }
+            }
+        }
+        if (message instanceof ZonePartitionReply) {
+            ZonePartitionReply reply = (ZonePartitionReply) message;
+            for (int i = 0; i < ElkMessageFactory.MAX_ZONES; i++) {
+                Thing thing = getThingForType(ElkTypeToRequest.Area, reply.getAreas()[i]);
+                if (thing == null && !areas[reply.getAreas()[i] - 1]) {
```

Comment:
Need to bring up jberry's comment on the original PR (https://github.com/openhab/openhab2-addons/pull/1955/files/08fd158129411c7bc6d8b7c67e4512f5042857f9#r147011052) - need a zero check on reply.getAreas. This was what stopped my discovery from working, once I added jberry's suggestion Discovery worked.

Reply:
Thanks!